google-books-bot/google_books_search.py

```diff
@@ -0,0 +1,110 @@
+#! /usr/bin/env python3
+import argparse
+
+from apiclient.discovery import build
+from olclient.openlibrary import OpenLibrary
+import olclient.common as ol_common
+
+NUMBER_OF_BOOK_CHOICES = 10
+
+
+def _isbn_matches(ol_book, isbn):
+    return [isbn] in ol_book.identifiers.values()
+
+
+def _ol_identifiers_from_google_identifiers(google_identifiers):
+    identifiers = {identifier["type"].lower(): [identifier["identifier"]] for identifier in google_identifiers}
+    if "isbn_10" not in identifiers and "isbn_13" not in identifiers:
+        raise KeyError
+    return identifiers
+
+
+def _ol_book_from_google_book(google_book):
+    google_book_info = google_book["volumeInfo"]
+
+    # Extract fields that we expect to always exist
+    title = google_book_info["title"]
+    authors = [ol_common.Author(name=author_name) for author_name in google_book_info["authors"]]
+    industry_identifiers = _ol_identifiers_from_google_identifiers(google_book_info["industryIdentifiers"])
+
+    # Extract fields that may not exist
+    number_of_pages = google_book_info.get("pageCount")
+    publisher = google_book_info.get("publisher")
+    publish_date = google_book_info.get("publishedDate")
+    cover_url = google_book_info.get("imageLinks", {}).get("thumbnail")
+
+    return ol_common.Book(title=title,
+                          number_of_pages=number_of_pages,
+                          identifiers=industry_identifiers,
+                          authors=authors,
+                          publisher=publisher,
+                          publish_date=publish_date,
+                          cover=cover_url)
+
+
+def _ol_books_from_google_books(google_books, max_books):
+    ol_books = []
+    for google_book in google_books:
+        try:
+            ol_books.append(_ol_book_from_google_book(google_book))
+        except KeyError:
+            continue
+
+        if len(ol_books) == max_books:
+            break
+
+    return ol_books
+
+
+def _upload_ol_book(ol_book):
+    existing_book = OL.Work.search(title=ol_book.title, author=ol_book.primary_author.name)
+    if existing_book is not None:
+        raise ValueError("It looks like this book already exists on Open Library. "
+                         "This script doesn't yet support updating existing books -- sorry!")
+
+    edition = OL.Work.create(ol_book)
+    print("Upload of {} successful!".format(edition.olid))
+
+
+def main():
+    global OL
```

Comment:
why not pass ol, rather than have a `global`? This would make (e.g.) test cases a lot easier to implement.

Reply:
I actually wrote this that way first, but rewrote to be more consistent with https://github.com/internetarchive/openlibrary-bots/blob/master/NY-Times-Bestseller-Bot/nyt_bestsellers_bot.py#L211 since this isn't my codebase ;)

Will fix.

Reply:
will submit a hotfix on `global OL` so we can merge and prevent from getting too stale.